server/src/lib/model/db/migrations/20200603213417-add-demographics-table.ts

```diff
@@ -0,0 +1,37 @@
+export const up = async function (db: any): Promise<any> {
+  return db.runSql(
+    `
+      CREATE TABLE demographics (
+        id BIGINT(20) UNSIGNED NOT NULL AUTO_INCREMENT PRIMARY KEY,
+        client_id CHAR(36) NOT NULL,
+        age VARCHAR(255) DEFAULT NULL,
+        gender VARCHAR(255) DEFAULT NULL,
+        updated_at DATETIME DEFAULT now(),
+        FOREIGN KEY (client_id) REFERENCES user_clients (client_id),
+        UNIQUE (client_id, age, gender)
+      );
+
+      CREATE TABLE user_client_demographics (
+        client_id BIGINT(20) UNSIGNED NOT NULL AUTO_INCREMENT PRIMARY KEY,
+        demographic_id BIGINT(20) UNSIGNED NOT NULL AUTO_INCREMENT,
+        FOREIGN KEY (client_id) REFERENCES user_clients (client_id),
+        FOREIGN KEY (demographic_id) REFERENCES demographics (id)
+      )
```

Comment:
Totally agreed about not running a big backfill!

Re: keeping everything in one table - my understanding of SQL sort is that it only acts on the preceding result set, so unless we have a user who's updating their demographics millions of time the sort should be on < 100 entries. (Also, MySQL 8.0 supports `DESC` indices! We're currently still on 5.7 but specifying a `DESC` index is just ignored in 5.6/5.7 and we plan on upgrading to 8.0 in Q3, so you could pre-emptively add it to `demographics.updated_at` if you'd like.) I know joins are more efficient than subqueries, but since joins are matrix multiplication it feels like it'll be slower than sorting because the volume of data it has to examine is multiplicative? I also have not tested this though and could be totally offbase 😅 

I'm not against having two tables, but if we do stick to them could we name them something more self-explanatory? like `demographics_current` and `demographics_historical`? I just know I'm going to forget which is which in six months.

Reply:
These are some awesome points and I am fully in favour of dropping any confusing tables that we don't need. Will update the PR :)